{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Mapping the residue distribution of pathogenic missense variants in disordered regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "# Add the parent directory of \"notebooks\" to Python's search path\n",
    "sys.path.append(os.path.abspath(\"..\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Computing fraction of disordered and folded mutations assigned to each pathogenicity category\n",
    "\n",
    "disordered_pathogenic = 7907\n",
    "disordered_benign = 20099\n",
    "disordered_VUS = 2691081\n",
    "folded_pathogenic = 54434\n",
    "folded_benign = 79931\n",
    "folded_VUS = 4828800\n",
    "\n",
    "### import os\n",
    "### from protein import Protein\n",
    "### >>> for UniProt_ID in os.listdir('/neuhaus/eytan/DisVar_protein_data'):\n",
    "### ...     protein = Protein(file_path = os.path.join('/neuhaus/eytan/DisVar_protein_data', UniProt_ID))\n",
    "### ...     for variant in protein.missense_variants['disordered']:\n",
    "### ...             if protein.missense_variants['disordered'][variant] in ['benign', 'likely_benign']:\n",
    "### ...                     disordered_benign += 1\n",
    "### ...             elif protein.missense_variants['disordered'][variant] in ['pathogenic', 'likely pathogenic']:\n",
    "### ...                     disordered_pathogenic += 1\n",
    "### ...             else:\n",
    "### ...                     disordered_VUS += 1\n",
    "### ...     for variant in protein.missense_variants['folded']:\n",
    "### ...             if protein.missense_variants['folded'][variant] in ['benign', 'likely benign']:\n",
    "### ...                     folded_benign += 1\n",
    "### ...             elif protein.missense_variants['folded'][variant] in ['pathogenic', 'likely pathogenic']:\n",
    "### ...                     folded_pathogenic += 1\n",
    "### ...             else:\n",
    "### ...                     folded_VUS += 1\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "labels = [\"Pathogenic\", \"Benign\", \"VUS\"]\n",
    "counts = [7907, 20099, 2691081]\n",
    "total = sum(counts)\n",
    "label_texts = [\n",
    "    f\"{lbl}: n = {cnt} ({cnt/total*100:.4f}%)\"\n",
    "    for lbl, cnt in zip(labels, counts)\n",
    "]\n",
    "fig, ax = plt.subplots(figsize=(10, 7))\n",
    "explode = [0.12, 0.12, 0.02]\n",
    "wedges, _ = ax.pie(\n",
    "    counts,\n",
    "    explode=explode,\n",
    "    startangle=90\n",
    ")\n",
    "right_y_positions = np.linspace(0.8, -0.8, 3)\n",
    "left_y_positions = np.linspace(0.8, -0.8, 3)\n",
    "right_i = 0\n",
    "left_i = 0\n",
    "for wedge, text in zip(wedges, label_texts):\n",
    "    angle = 0.5 * (wedge.theta1 + wedge.theta2)\n",
    "    rad = np.deg2rad(angle)\n",
    "    x = np.cos(rad)\n",
    "    y = np.sin(rad)\n",
    "    right_side = -90 <= angle <= 90\n",
    "    if right_side:\n",
    "        y_text = right_y_positions[right_i]\n",
    "        x_text = 1.6\n",
    "        ha = \"left\"\n",
    "        right_i += 1\n",
    "    else:\n",
    "        y_text = left_y_positions[left_i]\n",
    "        x_text = -1.6\n",
    "        ha = \"right\"\n",
    "        left_i += 1\n",
    "    ax.annotate(\n",
    "        text,\n",
    "        xy=(x, y),\n",
    "        xytext=(x_text, y_text),\n",
    "        ha=ha,\n",
    "        va=\"center\",\n",
    "        fontsize=12,\n",
    "        arrowprops=dict(arrowstyle=\"-\", lw=1.2)\n",
    "    )\n",
    "plt.title(\"Pathogenicity Annotations for Disordered Variants\", fontsize=16)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "labels = [\"Pathogenic\", \"Benign\", \"VUS\"]\n",
    "counts = [54434, 79931, 4828800]\n",
    "total = sum(counts)\n",
    "label_texts = [\n",
    "    f\"{lbl}: n = {cnt} ({cnt/total*100:.4f}%)\"\n",
    "    for lbl, cnt in zip(labels, counts)\n",
    "]\n",
    "fig, ax = plt.subplots(figsize=(10, 7))\n",
    "explode = [0.12, 0.12, 0.02]\n",
    "wedges, _ = ax.pie(\n",
    "    counts,\n",
    "    explode=explode,\n",
    "    startangle=90\n",
    ")\n",
    "right_y_positions = np.linspace(0.8, -0.8, 3)\n",
    "left_y_positions = np.linspace(0.8, -0.8, 3)\n",
    "right_i = 0\n",
    "left_i = 0\n",
    "for wedge, text in zip(wedges, label_texts):\n",
    "    angle = 0.5 * (wedge.theta1 + wedge.theta2)\n",
    "    rad = np.deg2rad(angle)\n",
    "    x = np.cos(rad)\n",
    "    y = np.sin(rad)\n",
    "    right_side = -90 <= angle <= 90\n",
    "    if right_side:\n",
    "        y_text = right_y_positions[right_i]\n",
    "        x_text = 1.6\n",
    "        ha = \"left\"\n",
    "        right_i += 1\n",
    "    else:\n",
    "        y_text = left_y_positions[left_i]\n",
    "        x_text = -1.6\n",
    "        ha = \"right\"\n",
    "        left_i += 1\n",
    "    ax.annotate(\n",
    "        text,\n",
    "        xy=(x, y),\n",
    "        xytext=(x_text, y_text),\n",
    "        ha=ha,\n",
    "        va=\"center\",\n",
    "        fontsize=12,\n",
    "        arrowprops=dict(arrowstyle=\"-\", lw=1.2)\n",
    "    )\n",
    "plt.title(\"Pathogenicity Annotations for Folded Variants\", fontsize=16)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Computing residue abundance for disordered and folded regions\n",
    "\n",
    "### >>> import os\n",
    "### >>> proteome_dir = '/neuhaus/eytan/DisVar_protein_data'\n",
    "### >>> from protein import Protein\n",
    "### >>> for UniProt_ID in os.listdir(proteome_dir):\n",
    "### ...     protein = Protein(file_path = os.path.join(proteome_dir, UniProt_ID))\n",
    "### >>> disordered_aa = ''\n",
    "### >>> folded_aa = ''\n",
    "### >>> for UniProt_ID in os.listdir(proteome_dir):\n",
    "### ...     protein = Protein(file_path = os.path.join(proteome_dir, UniProt_ID))\n",
    "### ...     disordered_nt = [(start, end) for start, end in protein.disordered_regions]\n",
    "### ...     for start, end in disordered_nt:\n",
    "### ...             disordered_aa += protein.aa_sequence[start : end]\n",
    "### ...     prev_end = 0\n",
    "### ...     for start, end in disordered_nt:\n",
    "### ...             if start > prev_end:\n",
    "### ...                     folded_aa += protein.aa_sequence[prev_end : start]\n",
    "### ...             prev_end = end\n",
    "### ...     if prev_end < len(protein.aa_sequence):\n",
    "### ...             folded_aa += protein.aa_sequence[prev_end : ]\n",
    "### ... \n",
    "### >>> len(disordered_aa)\n",
    "### 3670428\n",
    "### >>> len(folded_aa)\n",
    "### 7249747\n",
    "### >>> from collections import Counter\n",
    "### >>> def letter_fractions(s):\n",
    "### ...     counts = Counter(s)\n",
    "### ...     total = len(s)\n",
    "### ...     return {char: count / total for char, count in counts.items()}\n",
    "### >>> disordered_aa_fractions = letter_fractions(disordered_aa) \n",
    "### >>> folded_aa_fractions = letter_fractions(folded_aa) \n",
    "\n",
    "disordered_aa_fractions = {'M': 0.019112212526713506, 'A': 0.0767125795683773, 'D': 0.04827856587842072, 'T': 0.05656370319755625, 'P': 0.09918461825160445, 'N': 0.03203713572368127, 'G': 0.0790733396759179, 'Q': 0.05049356641786734, 'V': 0.045750522827310604, 'F': 0.022157361484818665, 'L': 0.07644394604661908, 'C': 0.014754137664599333, 'I': 0.02588362991999843, 'H': 0.02317440908798647, 'R': 0.05811311378400557, 'Y': 0.015405015436891828, 'K': 0.05600382298740093, 'S': 0.11524214614753375, 'E': 0.07810179085381869, 'W': 0.007514382518877908}\n",
    "folded_aa_fractions = {'M': 0.022509475158236557, 'T': 0.0513022040631211, 'N': 0.03808105303536799, 'K': 0.058547422413499396, 'L': 0.11149064925989831, 'D': 0.047390205478894644, 'A': 0.06671819030374439, 'W': 0.014324499875650832, 'S': 0.0668904721778567, 'R': 0.05536220781221745, 'F': 0.04390842880448104, 'V': 0.066689499647367, 'Y': 0.03255534296576142, 'C': 0.027010597749135244, 'P': 0.04419519743240695, 'G': 0.05861445923561195, 'H': 0.027751037381028607, 'E': 0.06804720219891812, 'Q': 0.04620671590332739, 'I': 0.05240513910347492}\n",
    "aa_order = sorted(disordered_aa_fractions.keys())\n",
    "disordered_values = [disordered_aa_fractions[aa] for aa in aa_order]\n",
    "folded_values = [folded_aa_fractions[aa] for aa in aa_order]\n",
    "x = np.arange(len(aa_order))\n",
    "width = 0.35\n",
    "fig, ax = plt.subplots(figsize=(14,6))\n",
    "ax.bar(x - width/2, disordered_values, width, label='Disordered')\n",
    "ax.bar(x + width/2, folded_values, width, label='Folded')\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(aa_order)\n",
    "ax.set_xlabel('amino acid')\n",
    "ax.set_ylabel('frequency')\n",
    "ax.set_title('Amino Acid Frequency in Disordered vs Folded Regions')\n",
    "ax.legend()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DisVar",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
